Acid Component

TemporalOroboros opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

- Makes acid levels a component.
- Merges the acid effect object into the component.
- Reworks acids decay rates slightly.
- Rebalances xenos acid spit so that they can still melt through walls.

<details>
<summary>Misc. associated changes:</summary>

- Adds defines for a lot of the acid associated constants.
- Documents clean types and adds `CLEAN_TYPE_ACID`
- Adds and implements a return bitflag for `COMSIG_COMPONENT_CLEAN_ACT`
- Adds a looping sound for acid.
- Makes `/atom/proc/acid_act` return a boolean.
- Fixes waterclosets creating a new reagent holder datum every time they are used.
- Removes waterclosets regenerating reagents on-use and restricts their reaction volume to 5 units.
- Adds and implements a couple reagent signals.
- Renames a few vars so Rohesie can stop telling me to rename more vars.

</details>

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
- \+ Moves the acid handling code for objects and turfs into more-or-less the same place.
- \+ Acid now tracks acid power and acid volume separately.
- \+ If someone wanted to implement it they could have separate acid overlays for separate objects.
- \- The acid-dilution effect of water is a bit more snowflakey, I'll see if I can fix it.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
code: Acid levels have been componentized.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

actions-user committed:
Automatic changelog compile [ci skip]

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Acid Component

TemporalOroboros committed:
Merge branch 'master' into acid_component

Rohesie commented on code/controllers/subsystem/acid.dm:
```diff
@@ -13,25 +13,23 @@ SUBSYSTEM_DEF(acid)
 
-	while (currentrun.len)
-		var/obj/O = currentrun[currentrun.len]
+	while(currentrun.len)
```
```suggestion
	while(length(currentrun))
```

Rohesie commented on code/controllers/subsystem/acid.dm:
```diff
@@ -13,25 +13,23 @@ SUBSYSTEM_DEF(acid)
-	while (currentrun.len)
-		var/obj/O = currentrun[currentrun.len]
+	while(currentrun.len)
+		var/datum/component/acid/acid = currentrun[currentrun.len]
```
```suggestion
		var/datum/component/acid/acid = currentrun[length(currentrun)]
```

Rohesie commented on code/controllers/subsystem/acid.dm:
```diff
@@ -13,25 +13,23 @@ SUBSYSTEM_DEF(acid)
-		if (!O || QDELETED(O))
-			processing -= O
-			if (MC_TICK_CHECK)
+		if(!acid || QDELETED(acid))
```
```suggestion
		if(QDELETED(acid))
```

Rohesie commented on html/changelog.html:
```diff
@@ -51,6 +51,20 @@
 		
 		-->
 		<div class="commit sansserif">
+			<h2 class="date">24 August 2020</h2>
+			<h3 class="author">Jared-Fogle updated:</h3>
+			<ul class="changes bgimages16">
+				<li class="bugfix">Being polled as a ghost for things like imaginary friends will now show the player's real name, rather than their displayed name.</li>
+				<li class="bugfix">You can now sign your name on paper again. Remember: %s.</li>
+			</ul>
+			<h3 class="author">tralezab updated:</h3>
+			<ul class="changes bgimages16">
+				<li class="tweak">when admins change the emergency shuttle timer, it gives a way more accurate estimate.</li>
+			</ul>
+			<h3 class="author">uomo91 updated:</h3>
+			<ul class="changes bgimages16">
+				<li class="tweak">Toolbelts can now hold pipe painters.</li>
+				<li class="bugfix">Bookshelves do not show shelves inside anymore when you crowbar them out.</li>
```
Disable actions in your repos to avoid this kind of unwanted changes by the bots.

TemporalOroboros replied:
I thought I already did. I'll check again.

TemporalOroboros replied:
I think I webedited this change out of existance.

Rohesie replied:
This is still a thing. Check the very first commit in this PR. Revert it.

TemporalOroboros replied:
Goddamnit

Rohesie commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -162,18 +162,20 @@
-	if(A)
-		A.acid_level = max(A.acid_level - reac_volume*50, 0)
+
+	// TODO: Desnowflake this.
+	var/datum/component/acid/turf_acid = T.GetComponent(/datum/component/acid)
```
Avoid the usage of `GetComponent`. What you want to do here is to send a signal at the start of the proc, and have the component registered to it.

TemporalOroboros replied:
I've done it and it does _not_ look pretty.

TemporalOroboros replied:
If you have any idea on how to do it better I'd live to hear it.

TemporalOroboros replied:
S C O P E   C R E E P

Rohesie commented on code/game/atoms.dm:
```diff
@@ -74,6 +74,9 @@
 	/// Radiation insulation types
 	var/rad_insulation = RAD_NO_INSULATION
 
+	/// The icon state intended to be used for the acid component. Used to override the default acid overlay icon state.
+	var/custom_acid_overlay = null
```
Can't this be stored on the component?

Qustinnus replied:
there's some issues with doing that because an object doesn't apply acid itself, it gets acid applied to it, the only other way of doing this is having a list that keeps the icons listed in an assoc list with type as key but honestly thats not great either.

Rohesie replied:
Yea, that'd offer no gains.
I suppose that short of an acidable element this is fine.

Rohesie commented on code/datums/components/acid.dm:
```diff
@@ -0,0 +1,178 @@
+	var/parent_integrity = 30
+
+/datum/component/acid/Initialize(_acid_power, _acid_volume)
+	if((_acid_power) <= 0 || (_acid_volume <= 0))
+		return COMPONENT_INCOMPATIBLE // Not enough acid.
```
Add an informative `stack_trace` message here. This shouldn't happen by default, should it?

TemporalOroboros replied:
This should not happen be default, no. I'll make some changes.

TemporalOroboros replied:
I've added `WARNING` calls and have prevented the `AddComponent` calls from running under these circumstances. That should be enough.

Rohesie commented on code/datums/components/acid.dm:
```diff
@@ -0,0 +1,178 @@
+	if(!isatom(parent))
+		return COMPONENT_INCOMPATIBLE // Incompatible type. TODO: Rework take_damage to the atom level and move this there.
+	if(isobj(parent))
+		var/obj/O = parent
+		if(O.resistance_flags & UNACIDABLE) // The parent object cannot have acid.
```
```suggestion
		var/obj/obj_parent = parent
		if(obj_parent.resistance_flags & UNACIDABLE) // The parent object cannot have acid.
```
Prefer better variable names.

Rohesie commented on code/datums/components/acid.dm:
```diff
@@ -0,0 +1,178 @@
+	var/atom/parent_atom = parent
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_OVERLAYS, .proc/on_update_overlays)
+	parent_atom.update_icon()
+	sizzle = new/datum/looping_sound/acid(list(parent), TRUE)
```
```suggestion
	sizzle = new /datum/looping_sound/acid(list(parent), TRUE)
```

Qustinnus replied:
the sound path should be in the var definition and this should just be sizzle = new sizzle(list(parent), TRUE)

Rohesie commented on code/datums/components/acid.dm:
```diff
@@ -0,0 +1,178 @@
+
+/datum/component/acid/Destroy(force, silent)
+	STOP_PROCESSING(SSacid, src)
+	qdel(sizzle)
```
```suggestion
	QDEL_NULL(sizzle)
```
Avoid keeping references, else you'll get a hard delete.

Rohesie commented on code/datums/components/acid.dm:
```diff
@@ -0,0 +1,178 @@
+	STOP_PROCESSING(SSacid, src)
+	qdel(sizzle)
+	UnregisterSignal(parent, COMSIG_ATOM_UPDATE_OVERLAYS)
+	if(parent && !QDELETED(parent))
```
```suggestion
	if(parent && !QDELING(parent))
```

Rohesie commented on code/datums/components/acid.dm:
```diff
@@ -0,0 +1,178 @@
+		RegisterSignal(parent, COMSIG_MOVABLE_CROSSED, .proc/on_crossed)
+
+/datum/component/acid/UnregisterFromParent()
+	UnregisterSignal(parent, list(	\
+		COMSIG_PARENT_EXAMINE,		\
+		COMSIG_COMPONENT_CLEAN_ACT,	\
+		COMSIG_ATOM_ATTACK_HAND		))
```
```suggestion
	UnregisterSignal(parent, list(
		COMSIG_PARENT_EXAMINE,
		COMSIG_COMPONENT_CLEAN_ACT,
		COMSIG_ATOM_ATTACK_HAND,
		))
```

Rohesie commented on code/datums/components/acid.dm:
```diff
@@ -0,0 +1,178 @@
+
+/// Sets the acid volume to a new value. Limits the acid volume by the amount allowed to exist on the parent atom.
+/datum/component/acid/proc/set_volume(new_volume)
+	if(isobj(parent))
+		new_volume = clamp(new_volume, 0, OBJ_ACID_VOLUME_MAX)
+	else if(isturf(parent))
```
Instead of checking types each time, store the difference of behaviors in variables, and only check for type on init.

Rohesie commented on code/datums/components/acid.dm:
```diff
@@ -0,0 +1,178 @@
+
+/// Handles the slow corrosion of the parent object.
+/datum/component/acid/process()
+	if(isobj(parent))
+		var/obj/O = parent
+		if(!(O.resistance_flags & ACID_PROOF))
+			O.take_damage(min(1 + round(sqrt(acid_power * acid_volume)*0.3), OBJ_ACID_DAMAGE_MAX), BURN, ACID, 0)
+
+	else if(isturf(parent))
```
Same here. You can use callbacks to store different behaviors instead of typechecking each time.

TemporalOroboros replied:
Should be done.

Qustinnus reviewed (changes requested):
The acid subsystem seems like its mostly just a copypasta of what processing.dm does. Could you instead move the acid subsystem into the processing folder and use PROCESSING_SUBSYSTEM_DEF instead? It means you can kill off a lot of duplicate code in your fire() proc.

TemporalOroboros committed:
Apply suggestions from code review

Mostly Rohesie's work.

Co-authored-by: Rohesie <rohesie@gmail.com>

TemporalOroboros committed:
Requested Changes

- Rohesie:
  - Callbacks for processing.
  - No creating acid components with <= zero power or volume.
  - Determines max_volume on initialize.
  - Verbose vars.
- Qustinnus
  - Moved SSacid to a processing subsystem.

Assorted syntax.

TemporalOroboros committed:
Merge branch 'acid_component' of https://github.com/TemporalOroboros/tgstation into acid_component

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Comment

// Blah

TemporalOroboros commented:
> The acid subsystem seems like its mostly just a copypasta of what processing.dm does. Could you instead move the acid subsystem into the processing folder and use PROCESSING_SUBSYSTEM_DEF instead? It means you can kill off a lot of duplicate code in your fire() proc.

Done, but I can't resolve your post.

Qustinnus commented:
its a review so it doesnt work the same, but its fine :)

TemporalOroboros committed:
Dilution

Adds a _very_ jank signal. If anyone has any better way of doing this, please tell me.

TemporalOroboros committed:
Merge branch 'master' into acid_component

TemporalOroboros committed:
Appveyor, show me the target of your rage.

TemporalOroboros committed:
Should get rid of the changelog changes.

Rohesie commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -389,7 +389,8 @@
 		busy = FALSE
 		O.wash(CLEAN_WASH)
-		O.acid_level = 0
+		create_reagents(5)
+		reagents.add_reagent(dispensedreagent, 5)
```
Why is it creating a new reagent datum every wash?

TemporalOroboros replied:
I don't know. I don't know why that change is there either. I'll look into it.

Rohesie commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -147,6 +147,10 @@
 		return
+
+	// This is almost certainly not the correct usage of signals and very much should be generalized.
+	SEND_SIGNAL(T, COMSIG_ACID_DILUTE, 0, reac_volume) // TODO: Make a robust pH system so that this can be handled generally.
```
Add this to the base: `/datum/reagent/proc/expose_turf(turf/T, volume)`
And then call parent here.

TemporalOroboros replied:
That will extend the effect to all reagents. Not just water as is the current behavior. Is this acceptable?

TemporalOroboros replied:
Time for scope creep.

TemporalOroboros replied:
I have applied scope creep. I think it worked.

Rohesie commented on code/datums/components/acid.dm:
```diff
@@ -0,0 +1,206 @@
+
+/datum/component/acid/Initialize(_acid_power, _acid_volume, _max_volume=null)
+	if((_acid_power) <= 0 || (_acid_volume <= 0))
+		WARNING("Acid component added with insufficient acid power ([_acid_power]) or acid volume ([_acid_power]).")
+		return COMPONENT_INCOMPATIBLE // Not enough acid.
+	if(!isatom(parent))
+		WARNING("Acid component added to [parent] ([parent?.type]) which is not a /atom subtype.")
+		return COMPONENT_INCOMPATIBLE // Incompatible type. TODO: Rework take_damage to the atom level and move this there.
+
+	if(isobj(parent))
+		var/obj/parent_object = parent
+		if(parent_object.resistance_flags & UNACIDABLE) // The parent object cannot have acid.
+			WARNING("Acid component added to unacidable object [parent].")
```
```suggestion
		stack_trace("Acid component added with insufficient acid power ([_acid_power]) or acid volume ([_acid_power]).")
		return COMPONENT_INCOMPATIBLE // Not enough acid.
	if(!isatom(parent))
		stack_trace("Acid component added to [parent] ([parent?.type]) which is not a /atom subtype.")
		return COMPONENT_INCOMPATIBLE // Incompatible type. TODO: Rework take_damage to the atom level and move this there.

	if(isobj(parent))
		var/obj/parent_object = parent
		if(parent_object.resistance_flags & UNACIDABLE) // The parent object cannot have acid.
			stack_trace("Acid component added to unacidable object [parent].")
```

TemporalOroboros replied:
Done, should be in the next push.

Rohesie commented on code/datums/components/acid.dm:
```diff
@@ -0,0 +1,206 @@
+
+/// Handles processing on a [/turf].
+/datum/component/acid/proc/process_turf(turf/target)
+	var/turf/T = parent
+	var/acid_used = min(acid_volume * 0.05, 20)
+	var/applied_targets = 0
+	for(var/A in T)
+		var/atom/acid_target = A
```
```suggestion
	var/turf/parent_turf = parent
	var/acid_used = min(acid_volume * 0.05, 20)
	var/applied_targets = 0
	for(var/am in parent_turf)
		var/atom/movable/acid_target = am
```

TemporalOroboros replied:
Done, should be in the next push.

Rohesie commented on code/datums/components/acid.dm:
```diff
@@ -0,0 +1,206 @@
+	if(!(clean_types & CLEAN_TYPE_ACID))
+		return NONE
+	qdel(src)
+	return TRUE // I know that returning booleans is bad form but this is how the proc handles it.
```
Don't, create a flag. Check in the signal definitions for examples.

TemporalOroboros replied:
Done, should be in the next push.

Rohesie replied:
If this proc is used elsewhere, not for signal returns, then make the signal handler call it and return a flag depending on its result.

TemporalOroboros replied:
I meant that the proc that sends the signal handled it like a boolean.

TemporalOroboros committed:
Requested Changes

- Rohesie
  - Adds and implements cleanable return bitflags.
  - WARNING -> stack_trace
  - Verbose vars
  - removed sinks auto-regenerating 5u every time they are used.
- acid_act now returns a boolean.

TemporalOroboros committed:
S C O P E   C R E E P

- Adds and implements a couple reagent signals.
- Preemptively verbosifies a couple vars so that Rohesie doesn't tell me to go back and fix them.

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into acid_component

TemporalOroboros committed:
Compile Warning Fix

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into acid_component

Rohesie commented on code/datums/components/radioactive.dm:
```diff
@@ -99,18 +99,19 @@
 
 	if(!(clean_types & CLEAN_TYPE_WEAK))
 		qdel(src)
-		return
+		return NONE
```
Stuff that delete source should probably return a value to stop the proc, no?

TemporalOroboros replied:
Done

Rohesie commented on code/datums/components/radioactive.dm:
```diff
@@ -99,18 +99,19 @@
 
 	strength = max(0, (strength - (RAD_BACKGROUND_RADIATION * 2)))
 	if(strength <= RAD_BACKGROUND_RADIATION)
 		qdel(src)
+		return NONE
```
Same here.

TemporalOroboros replied:
Done

Rohesie commented on code/modules/antagonists/blob/blobstrains/_reagent.dm:
```diff
@@ -27,7 +27,8 @@
 
-/datum/reagent/blob/expose_mob(mob/living/M, methods=TOUCH, reac_volume, show_message, touch_protection, mob/camera/blob/O)
-	if(M.stat == DEAD || istype(M, /mob/living/simple_animal/hostile/blob))
+/datum/reagent/blob/expose_mob(mob/living/exposed_mob, methods=TOUCH, reac_volume, show_message, touch_protection, mob/camera/blob/overmind)
+	SEND_SIGNAL(src, COMSIG_REAGENT_EXPOSE_MOB, exposed_mob, methods, reac_volume, show_message, touch_protection, overmind)
```
Signals can't have multiple sources. This is a bit of a mess, redefining the proc and not calling parent.
I don't want to creep the scope of this, but this needs to be refactored.

TemporalOroboros replied:
Yeah, it is a massive mess. I'll see what I can do.

TemporalOroboros replied:
I think I'll need to atomize the fix for this out. I'll see what I can do.

TemporalOroboros replied:
Never mind, using a set of bitflags. Reagent traits can be done another day.

Rohesie commented on code/modules/antagonists/blob/blobstrains/explosive_lattice.dm:
```diff
@@ -31,18 +31,18 @@
-			L.apply_damage(0.4*aoe_volume, BRUTE, wound_bonus=CANT_WOUND)
-		if(M)
-			M.apply_damage(0.6*reac_volume, BRUTE, wound_bonus=CANT_WOUND)
+			var/aoe_volume = ..(nearby_mob, TOUCH, initial_volume, 0, nearby_mob.get_permeability_protection(), overmind)
```
```suggestion
			exposed_mob = nearby_mob
			methods = TOUCH
			reac_volume = initial_volume
			show_message = FALSE
			touch_protection = nearby_mob.get_permeability_protection()
			var/aoe_volume = ..()
```

Rohesie commented on code/modules/antagonists/blob/blobstrains/pressurized_slime.dm:
```diff
@@ -39,14 +39,14 @@
+		location_turf.MakeSlippery(TURF_WET_LUBE, min_wet_time = 10 SECONDS, wet_time_to_add = 5 SECONDS)
+		exposed_mob.adjust_fire_stacks(-(reac_volume / 10))
+	exposed_mob.apply_damage(0.4*reac_volume, BRUTE, wound_bonus=CANT_WOUND)
+	if(exposed_mob)
+		exposed_mob.apply_damage(0.4*reac_volume, OXY)
+	if(exposed_mob)
+		exposed_mob.adjustStaminaLoss(reac_volume)
```
```suggestion
	if(exposed_mob)
		exposed_mob.adjustStaminaLoss(reac_volume, FALSE)
		exposed_mob.apply_damage(0.4 * reac_volume, OXY)
```

Rohesie commented on code/modules/reagents/chemistry/reagents/food_reagents.dm:
```diff
@@ -349,18 +356,19 @@
 		return
-	if(M.has_bane(BANE_SALT))
-		M.mind.disrupt_spells(-200)
+	SEND_SIGNAL(src, COMSIG_REAGENT_EXPOSE_MOB, exposed_mob, methods, reac_volume, show_message, touch_protection)
```
This should be done at the base of the proc.

TemporalOroboros replied:
Done

Rohesie commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -105,13 +107,16 @@
-				D.cure()
-		LAZYOR(L.disease_resistances, data)
+/datum/reagent/vaccine/expose_mob(mob/living/exposed_mob, methods=TOUCH, reac_volume, show_message=TRUE, touch_protection=0)
+	SEND_SIGNAL(src, COMSIG_REAGENT_EXPOSE_MOB, exposed_mob, methods, reac_volume, show_message, touch_protection)
```
Needs to be done at the base of it.

TemporalOroboros replied:
Done

Rohesie commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -689,9 +692,10 @@
 
-/datum/reagent/aslimetoxin/expose_mob(mob/living/L, methods=TOUCH, reac_volume)
+/datum/reagent/aslimetoxin/expose_mob(mob/living/exposed_mob, methods=TOUCH, reac_volume, show_message=TRUE, touch_protection=0)
+	SEND_SIGNAL(src, COMSIG_REAGENT_EXPOSE_MOB, exposed_mob, methods, reac_volume, show_message, touch_protection)
```
At the base of the proc.

TemporalOroboros replied:
Done

Rohesie commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -700,8 +704,9 @@
-/datum/reagent/gluttonytoxin/expose_mob(mob/living/L, methods=TOUCH, reac_volume)
-	L.ForceContractDisease(new /datum/disease/transformation/morph(), FALSE, TRUE)
+/datum/reagent/gluttonytoxin/expose_mob(mob/living/exposed_mob, methods=TOUCH, reac_volume, show_message=TRUE, touch_protection=0)
+	SEND_SIGNAL(src, COMSIG_REAGENT_EXPOSE_MOB, exposed_mob, methods, reac_volume, show_message, touch_protection)
```
At the base of the proc.

TemporalOroboros replied:
Done

Rohesie commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -1069,24 +1085,29 @@
 
-/datum/reagent/space_cleaner/expose_mob(mob/living/M, methods=TOUCH, reac_volume)
+/datum/reagent/space_cleaner/expose_mob(mob/living/exposed_mob, methods=TOUCH, reac_volume, show_message=TRUE, touch_protection=0)
+	SEND_SIGNAL(src, COMSIG_REAGENT_EXPOSE_MOB, exposed_mob, methods, reac_volume, show_message, touch_protection)
```
At the base of the proc.

TemporalOroboros replied:
Done

Rohesie commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -1825,13 +1860,16 @@
-		exposed_human.facial_hairstyle = "Shaved"
-		exposed_human.update_hair()
+/datum/reagent/baldium/expose_mob(mob/living/exposed_mob, methods=TOUCH, reac_volume, show_message=TRUE, touch_protection=FALSE)
+	SEND_SIGNAL(src, COMSIG_REAGENT_EXPOSE_MOB, exposed_mob, methods, reac_volume, show_message, touch_protection)
```
Should be done at the base of the proc.

TemporalOroboros replied:
Done

Rohesie commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -1809,14 +1842,16 @@
-			H.facial_hairstyle = "Beard (Very Long)"
-			H.update_hair()
+/datum/reagent/concentrated_barbers_aid/expose_mob(mob/living/exposed_mob, methods=TOUCH, reac_volume, show_message=TRUE, touch_protection=FALSE)
+	SEND_SIGNAL(src, COMSIG_REAGENT_EXPOSE_MOB, exposed_mob, methods, reac_volume, show_message, touch_protection)
```
Should be done at the base of the proc.

TemporalOroboros replied:
Done

Rohesie commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -1792,15 +1822,18 @@
-		exposed_human.facial_hairstyle = picked_beard
-		exposed_human.update_hair()
+/datum/reagent/barbers_aid/expose_mob(mob/living/exposed_mob, methods=TOUCH, reac_volume, show_message=TRUE, touch_protection=FALSE)
+	SEND_SIGNAL(src, COMSIG_REAGENT_EXPOSE_MOB, exposed_mob, methods, reac_volume, show_message, touch_protection)
```
Should be done at the base of the proc.

TemporalOroboros replied:
Done

Rohesie commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -1777,13 +1805,15 @@
-			H.facial_hair_color = pick(potential_colors)
-			H.update_hair()
+/datum/reagent/hair_dye/expose_mob(mob/living/exposed_mob, methods=TOUCH, reac_volume, show_message=TRUE, touch_protection=FALSE)
+	SEND_SIGNAL(src, COMSIG_REAGENT_EXPOSE_MOB, exposed_mob, methods, reac_volume, show_message, touch_protection)
```
Should be done at the base of the proc.

TemporalOroboros replied:
Done

Rohesie commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -1174,9 +1197,10 @@
 
-/datum/reagent/snail/expose_mob(mob/living/L, methods=TOUCH, reac_volume, show_message = 1, touch_protection = 0)
+/datum/reagent/snail/expose_mob(mob/living/exposed_mob, methods=TOUCH, reac_volume, show_message = TRUE, touch_protection = 0)
+	SEND_SIGNAL(src, COMSIG_REAGENT_EXPOSE_MOB, exposed_mob, methods, reac_volume, show_message, touch_protection)
```
Should be done at the base of the proc.

TemporalOroboros replied:
Done

Rohesie commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -1163,9 +1185,10 @@
 
-/datum/reagent/fungalspores/expose_mob(mob/living/L, methods=TOUCH, reac_volume, show_message = 1, touch_protection = 0)
+/datum/reagent/fungalspores/expose_mob(mob/living/exposed_mob, methods=TOUCH, reac_volume, show_message = TRUE, touch_protection = 0)
+	SEND_SIGNAL(src, COMSIG_REAGENT_EXPOSE_MOB, exposed_mob, methods, reac_volume, show_message, touch_protection)
```
Should be done at the base of the proc.

TemporalOroboros replied:
Done.

TemporalOroboros committed:
Apply suggestions from code review

Co-authored-by: Rohesie <rohesie@gmail.com>

TemporalOroboros committed:
SHOULD_CALL_PARENT

Makes reagent exposure procs SHOULD CALL_PARENT

Removes several signal calls.

TemporalOroboros committed:
Merge branch 'acid_component' of https://github.com/TemporalOroboros/tgstation into acid_component

TemporalOroboros committed:
Revert

TemporalOroboros committed:
Reverts bot action

TemporalOroboros committed:
Merge branch 'master' into acid_component

TemporalOroboros committed:
Revert "Reverts bot action"

This reverts commit 0c4f4cd396ba276882580d7de9362309033ef04b.

TemporalOroboros committed:
Merge branch 'master' into acid_component

TemporalOroboros committed:
Revert "Revert "Reverts bot action""

This reverts commit 35e19f80c8849b4e337ce11cc69020bfb61bacdb.

TemporalOroboros committed:
Revert "Revert "Revert "Reverts bot action"""

This reverts commit d90abdc48758d8b9369b221cfad82f2d802c3939.

TemporalOroboros committed:
Can this just finally die please?

TemporalOroboros committed:
Revert "Can this just finally die please?"

This reverts commit 413238b7d037ed799135a45d0ea94b279f61c76c.

TemporalOroboros committed:
Please fix

TemporalOroboros committed:
Revert "Please fix"

This reverts commit 9d0aea51ce3f596d6daee1ada64886158d638d3a.

TemporalOroboros committed:
Revert "Revert "Can this just finally die please?""

This reverts commit 9ad7b3a298c962327cc823ebe2b74d38ded026b6.

TemporalOroboros committed:
Revert "Can this just finally die please?"

This reverts commit 413238b7d037ed799135a45d0ea94b279f61c76c.

TemporalOroboros committed:
Revert "Revert "Revert "Revert "Reverts bot action""""

This reverts commit e0661238fb44b906de327a59b0ecb68f435c234a.

TemporalOroboros committed:
Revert "Revert "Revert "Reverts bot action"""

This reverts commit d90abdc48758d8b9369b221cfad82f2d802c3939.

TemporalOroboros committed:
Revert "Merge branch 'master' into acid_component"

This reverts commit 879a18c4437345ba8df99bf1c4f5885c250e23e2, reversing
changes made to e88e2cbdebc65e1f8edd8aed7b4441a3b7101fe4.

TemporalOroboros committed:
Revert "Reverts bot action"

This reverts commit 0c4f4cd396ba276882580d7de9362309033ef04b.

TemporalOroboros committed:
Revert "Merge branch 'master' into acid_component"

This reverts commit 6f05a8aa7c0af0e55e584c35823ee3cd4a7cf2ea, reversing
changes made to 088667ae023b92d6ead642095b0bcfd8a0912746.

TemporalOroboros committed:
Revert "Revert "Merge branch 'master' into acid_component""

This reverts commit 68beb44195343333cc588cfc954d9f8e68ee5338.

TemporalOroboros committed:
If this doesn't work I'm going to scream.

TemporalOroboros committed:
Revert "If this doesn't work I'm going to scream."

This reverts commit 9fd4a8fa64d79164983def0a6f2249cada17912b.

TemporalOroboros committed:
Revert "Revert "Please fix""

This reverts commit 293dda473745cc010a2fc2a2030a23c16a880524.

TemporalOroboros committed:
Let's hope that this fixes it.

TemporalOroboros committed:
Revert "Let's hope that this fixes it."

This reverts commit f0bfedfbde8d8cc72e2c25b51d3022f3962fd406.

TemporalOroboros committed:
Revert "Revert "Let's hope that this fixes it.""

This reverts commit d61c241fdcdad80779a8db1e3bbcb637e6c7db90.

Rohesie commented on html/changelog.html:
```diff
@@ -51,6 +51,32 @@
 		
 		-->
 		<div class="commit sansserif">
+			<h2 class="date">04 September 2020</h2>
+			<h3 class="author">Fikou updated:</h3>
+			<ul class="changes bgimages16">
+				<li class="tweak">spy glasses now use an action button instead of a verb, so they are less cryptic and someone will know how to use them</li>
+				<li class="balance">spy glasses are now sunglasses, removes the sunglasses from det's office so you can just take the spy ones from your locker</li>
+			</ul>
+			<h3 class="author">Ghilker updated:</h3>
+			<ul class="changes bgimages16">
+				<li class="bugfix">Icebox pipe layering error</li>
+			</ul>
+			<h3 class="author">SpaceManiac updated:</h3>
+			<ul class="changes bgimages16">
+				<li class="bugfix">Zipper noises are no longer played on the lobby screen when a character joins with a random-colored jumpsuit.</li>
+			</ul>
+			<h3 class="author">itseasytosee updated:</h3>
+			<ul class="changes bgimages16">
+				<li class="bugfix">supermatter tongs holding supermatter shards now have working in hand sprites</li>
+			</ul>
+			<h3 class="author">nightred updated:</h3>
+			<ul class="changes bgimages16">
+				<li class="tweak">The stomach is now needed to eat food or drink</li>
+				<li class="tweak">Vomiting now empties the stomach only,</li>
+				<li class="rscadd">New surgery to filter the blood of reagents and the tool that is needed</li>
+				<li class="tweak">Stomach pump surgery is now do able on the living</li>
+			</ul>
+
```
Still a thing. Disable your github actions and revert this commit.

TemporalOroboros replied:
I think it's finally dead. If it comes back the next time I merge master I don't know what I'll do.

TemporalOroboros replied:
It stayed gone. I think it's done.

Rohesie commented on code/modules/reagents/chemistry/reagents/toxin_reagents.dm:
```diff
@@ -38,20 +38,17 @@
-	if(!..())
-		return
-	if(!M.has_dna() || HAS_TRAIT(M, TRAIT_GENELESS) || HAS_TRAIT(M, TRAIT_BADDNA))
+/datum/reagent/toxin/mutagen/expose_mob(mob/living/carbon/exposed_mob, methods=TOUCH, reac_volume)
+	if(!(. = ..()) || !exposed_mob.has_dna() || HAS_TRAIT(exposed_mob, TRAIT_GENELESS) || HAS_TRAIT(exposed_mob, TRAIT_BADDNA))
```
```suggestion
/datum/reagent/toxin/mutagen/expose_mob(mob/living/carbon/exposed_mob, methods=TOUCH, reac_volume)
	. = ..()
	if(!. || !exposed_mob.has_dna() || HAS_TRAIT(exposed_mob, TRAIT_GENELESS) || HAS_TRAIT(exposed_mob, TRAIT_BADDNA))
```

Rohesie commented on code/modules/antagonists/blob/blobstrains/_reagent.dm:
```diff
@@ -25,9 +25,11 @@
-/datum/reagent/blob/expose_mob(mob/living/M, methods=TOUCH, reac_volume, show_message, touch_protection, mob/camera/blob/O)
-	if(M.stat == DEAD || istype(M, /mob/living/simple_animal/hostile/blob))
+/datum/reagent/blob/expose_mob(mob/living/exposed_mob, methods=TOUCH, reac_volume, show_message, touch_protection, mob/camera/blob/overmind)
+	..()
```
```suggestion
	. = ..()
```

Rohesie commented on code/datums/components/acid.dm:
```diff
@@ -0,0 +1,210 @@
+	sizzle = new(list(parent), TRUE)
+	START_PROCESSING(SSacid, src)
+
+/datum/component/acid/Destroy(force, silent)
+	STOP_PROCESSING(SSacid, src)
+	QDEL_NULL(sizzle)
+	UnregisterSignal(parent, COMSIG_ATOM_UPDATE_OVERLAYS)
```
```suggestion
/datum/component/acid/Destroy(force, silent)
	STOP_PROCESSING(SSacid, src)
	QDEL_NULL(sizzle)
	QDEL_NULL(process_effect)
	UnregisterSignal(parent, COMSIG_ATOM_UPDATE_OVERLAYS)
```

Rohesie commented:
This is a great refactor and cleanup, by the byes. Thanks for it.
On the changelog issue, you won't fix it by just reverting, you need to do that and then merge into master or rebase.

TemporalOroboros committed:
Apply suggestions from code review

Co-authored-by: Rohesie <rohesie@gmail.com>

TemporalOroboros committed:
Revert "Revert "Revert "Let's hope that this fixes it."""

This reverts commit e92882cf8d3b3c7f5318d387a305029d6e29ea52.

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'acid_component' of https://github.com/TemporalOroboros/tgstation into acid_component

TemporalOroboros committed:
Merge branch 'master' into acid_component

TemporalOroboros commented:
I reverted and merged the master. It's still there.

TemporalOroboros committed:
Here's hoping.

TemporalOroboros commented:
I think it's gone.

Rohesie commented on code/datums/components/acid.dm:
```diff
@@ -0,0 +1,211 @@
+
+/// Handles the slow corrosion of the parent [/atom].
+/datum/component/acid/process()
+	process_effect?.InvokeAsync()
+	set_volume(acid_volume - (ACID_DECAY_BASE + (ACID_DECAY_SCALING*round(sqrt(acid_volume)))))
```
`process_turf()` calls `set_volume()` and `process_obj()` calls `take_damage()`, meaning by the time it reaches this `set_volume()` call, `src` might be qdeleting, and this will make it try to qdelete twice.
```suggestion
	process_effect?.InvokeAsync()
	if(QDELETING(src))
		return
	set_volume(acid_volume - (ACID_DECAY_BASE + (ACID_DECAY_SCALING*round(sqrt(acid_volume)))))
```

TemporalOroboros committed:
Update code/datums/components/acid.dm

Co-authored-by: Rohesie <rohesie@gmail.com>

Rohesie commented on code/datums/components/acid.dm:
```diff
@@ -0,0 +1,213 @@
+/// Handles the slow corrosion of the parent [/atom].
+/datum/component/acid/process()
+	process_effect?.InvokeAsync()
+	if(QDELETING(src))
```
```suggestion
	if(QDELING(src)) //The process effect deals damage, and on turfs diminishes the acid volume, potentially destroying the component. Let's not destroy it twice.
```

TemporalOroboros committed:
Comments for appveyor.

Rohesie commented on code/datums/components/acid.dm:
```diff
@@ -0,0 +1,213 @@
+/datum/component/acid/Destroy(force, silent)
+	STOP_PROCESSING(SSacid, src)
+	QDEL_NULL(sizzle)
+	QDEL_NULL(process_effect)
```
```suggestion
	if(process_effect)
		QDEL_NULL(process_effect)
```

TemporalOroboros committed:
Merge branch 'acid_component' of https://github.com/TemporalOroboros/tgstation into acid_component

TemporalOroboros committed:
Apply suggestions from code review

Co-authored-by: Rohesie <rohesie@gmail.com>

TemporalOroboros committed:
QDEL

TemporalOroboros committed:
Merge branch 'acid_component' of https://github.com/TemporalOroboros/tgstation into acid_component

TemporalOroboros committed:
"""sizzle"""

Rohesie commented:
Missed the fact that several signal handlers are returning booleans, as highlighted by #53486
